### ISSUE TITLE
Yarn 1.x: Investigate replacing pyarn with the native Yarn yarn.lock parser

### ISSUE THREAD

**Issue by taylormadore** [containerbuildsystem/cachi2#716]
In our initial implementation for the Yarn 1.x package manager in cachi2, we will use the [pyarn](https://github.com/containerbuildsystem/pyarn) library for parsing yarn.lock. Pyarn is a library that we maintain and there may be some benefit to use yarn itself instead to parse yarn.lock.

Here is Yarn's own parser: https://github.com/yarnpkg/yarn/tree/master/packages/lockfile
It is also distributed as a [package](https://www.npmjs.com/package/@yarnpkg/lockfile) in the npm registry, but caution that it has not been released for many years even though some changes have been made to the lockfile code in the repository.

Additional context from the discussion on the design: https://github.com/containerbuildsystem/cachi2/pull/585#discussion_r1725173960

**Comment by derasdf** [comment]
After some investigation, I believe this is doable. However, we need to decide how we want to add the `@yarnpkg/lockfile` dependency from the npm [package](https://www.npmjs.com/package/@yarnpkg/lockfile) to the project. I'm open to suggestions on the best way of incorporating it.

**Comment by taylormadore** [comment]
Is the published `@yarnpkg/lockfile` consistent with the current code in the upstream yarn repo? It looks so old that it makes me suspect that they stopped publishing it at some point

**Comment by derasdf** [comment]
You're right, for some reason, the versions aren’t consistent. The version published on npm is `1.1.0`, while the one in the GitHub repo is `1.0.2`.

**Comment by slimreaper35** [comment]
Could we copy the source file from the yarn repo without installing it directly with `npm` ?
Similarly, what we have for bundler -> https://github.com/hermetoproject/hermeto/blob/main/hermeto/core/package_managers/bundler/scripts/lockfile_parser.rb

**Comment by eskultety** [comment]
> Could we copy the source file from the yarn repo without installing it directly with `npm` ? Similarly, what we have for bundler -> https://github.com/hermetoproject/hermeto/blob/main/hermeto/core/package_managers/bundler/scripts/lockfile_parser.rb

What source file? https://github.com/yarnpkg/yarn/tree/master/src/lockfile is a package depending on several modules, we'd have to copy potentially many sources, the only other alternative really to having to re-introduce a fake mini nodejs project inside our repo is to have the whole yarn repo as a git submodule IIUC because everything is hosted inside a single repository.

**Comment by eskultety** [comment]
> Is the published `@yarnpkg/lockfile` consistent with the current code in the upstream yarn repo? It looks so old that it makes me suspect that they stopped publishing it at some point

It's weirder than that, they have lockfile changes on the master branch that were never released and the 1.1.0 release doesn't correspond to any tagged release in the repo; backporting also doesn't sound like an explanation because the changes in git happened 5 years ago while the 1.1.0 release in the npm registry is from 7 years ago, so we'll have to diff the tarballs to truly see the differences having a complete lack of changelog.

**Comment by eskultety** [comment]
> > Is the published `@yarnpkg/lockfile` consistent with the current code in the upstream yarn repo? It looks so old that it makes me suspect that they stopped publishing it at some point
> 
> It's weirder than that, they have lockfile changes on the master branch that were never released and the 1.1.0 release doesn't correspond to any tagged release in the repo; backporting also doesn't sound like an explanation because the changes in git happened 5 years ago while the 1.1.0 release in the npm registry is from 7 years ago, so we'll have to diff the tarballs to truly see the differences having a complete lack of changelog.

Did a diff on the upstream release vs the latest version available from the NPM registry last week. Well, there's some downstream happening clearly which isn't linked anywhere from the package's npm registry page. There are a lot of changes, though most of them are just stylistic changes. Regardless, given the situation it seems like we're pretty much left out with consuming the package from the registry as the only viable option.
